service/controller/app/resource/chart/desired.go

```diff
@@ -49,21 +50,36 @@ func (r *Resource) GetDesiredState(ctx context.Context, obj interface{}) (interf
 		return nil, microerror.Mask(err)
 	}
 
-	index, err := r.indexCache.GetIndex(ctx, key.CatalogStorageURL(cc.Catalog))
-	if err != nil {
-		r.logger.Errorf(ctx, err, "failed to get index.yaml")
-	}
+	var tarballURL, version string
 
-	version, tarballURL, err := getVersionAndTarballURL(index, key.AppName(cr), cr.Spec.Version)
-	if err != nil {
-		r.logger.Errorf(ctx, err, "failed to get tarball URL")
-	}
+	if key.CatalogVisibility(cc.Catalog) == "internal" {
+		// For internal catalogs we generate the URL as its predicatable
+		// and to avoid having chicken egg problems.
+		tarballURL, err = appcatalog.NewTarballURL(key.CatalogStorageURL(cc.Catalog), key.AppName(cr), key.Version(cr))
+		if err != nil {
+			r.logger.Errorf(ctx, err, "failed to generated tarball")
+		}
 
-	if !isValidURL(tarballURL) {
-		// URL may be relative. If so we join it to the Catalog Storage URL.
-		tarballURL, err = joinRelativeURL(cc.Catalog, tarballURL)
+		version = key.Version(cr)
```

Comment:
This is what we did before v5.6.0. 

For our catalogs the URL is predictable for community catalogs it can be anything. So we need to check the index.yaml.